core/src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -716,6 +719,38 @@ public String toString() {
             return "store(" + in.toString() + ")";
         }
 
+        @Override
+        public void copyFrom(Directory from, String srcFile, String destFile, IOContext context) throws IOException {
+            Directory fromUnwrapped = FilterDirectory.unwrap(from);
+            Directory toUnwrapped = FilterDirectory.unwrap(this);
+            // try to unwrap to FSDirectory - we might be able to just create hard-links of these files and safe copying
+            // the entire file.
+            if (fromUnwrapped instanceof FSDirectory
+                && toUnwrapped instanceof FSDirectory) {
+                final Path fromPath = ((FSDirectory)fromUnwrapped).getDirectory();
+                final Path toPath = ((FSDirectory)toUnwrapped).getDirectory();
+                try {
+                    Files.createLink(toPath.resolve(destFile), fromPath.resolve(srcFile));
+                } catch (FileNotFoundException | NoSuchFileException | FileAlreadyExistsException ex) {
+                    throw ex; // in these cases we bubble up since it's a true error condition.
+                } catch (IOException
+                    | UnsupportedOperationException // if the FS doesn't support hard-links
+                    | SecurityException ex // we might miss a java.nio.file.LinkPermission" "hard"
```

Comment:
should we be defensive on this? I mean we explicitly allow for this permission - we might be hiding something else with this?